frame/contracts/src/chain_extension.rs

```diff
@@ -0,0 +1,393 @@
+	pub(crate) fn new(
+		runtime: &'a mut Runtime::<'b, E>,
+		input_ptr: u32,
+		input_len: u32,
+		output_ptr: u32,
+		output_len_ptr: u32,
+	) -> Self {
+		Environment {
+			inner: Inner {
+				runtime,
+				input_ptr,
+				input_len,
+				output_ptr,
+				output_len_ptr,
+			},
+			state: PhantomData,
+		}
+	}
+
+	/// Use all arguments as integer values.
+	pub fn only_in(self) -> Environment<'a, 'b, E, state::OnlyIn> {
+		Environment {
+			inner: self.inner,
+			state: PhantomData,
+		}
+	}
+
+	/// Use input arguments as integer and output arguments as pointer to a buffer.
+	pub fn prim_in_buf_out(self) -> Environment<'a, 'b, E, state::PrimInBufOut> {
+		Environment {
+			inner: self.inner,
+			state: PhantomData,
+		}
+	}
+
+	/// Use input and output arguments as pointers to a buffer.
+	pub fn buf_in_buf_out(self) -> Environment<'a, 'b, E, state::BufInBufOut> {
+		Environment {
+			inner: self.inner,
+			state: PhantomData,
+		}
+	}
+}
+
+/// Functions to use the input arguments as integers.
+impl<'a, 'b, E: Ext, S: state::PrimIn> Environment<'a, 'b, E, S> {
+	/// The `input_ptr` argument.
+	pub fn val0(&self) -> u32 {
+		self.inner.input_ptr
+	}
+
+	/// The `input_len` argument.
+	pub fn val1(&self) -> u32 {
+		self.inner.input_len
+	}
+}
+
+/// Functions to use the output arguments as integers.
+impl<'a, 'b, E: Ext, S: state::PrimOut> Environment<'a, 'b, E, S> {
+	/// The `output_ptr` argument.
+	pub fn val2(&self) -> u32 {
+		self.inner.output_ptr
+	}
+
+	/// The `output_len_ptr` argument.
+	pub fn val3(&self) -> u32 {
+		self.inner.output_len_ptr
+	}
+}
+
+/// Functions to use the input arguments as pointer to a buffer.
+impl<'a, 'b, E: Ext, S: state::BufIn> Environment<'a, 'b, E, S>
+where
+	<E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+{
+	/// Reads `min(max_len, in_len)` from contract memory.
+	///
+	/// This does **not** charge any weight. The caller must make sure that the an
+	/// appropriate amount of weight is charged **before** reading from contract memory.
+	/// The reason for that is that usually the costs for reading data and processing
+	/// said data cannot be separated in a benchmark. Therefore a chain extension would
+	/// charge the overall costs either using `max_len` (worst case approximation) or using
+	/// [`in_len()`](Self::in_len).
+	pub fn read(&self, max_len: u32) -> Result<Vec<u8>> {
+		self.inner.runtime.read_sandbox_memory(
+			self.inner.input_ptr,
+			self.inner.input_len.min(max_len),
+		)
+	}
+
+	/// Reads `min(buffer.len(), in_len) from contract memory.
+	///
+	/// This takes a mutable pointer to a buffer fills it with data and shrinks it to
+	/// the size of the actual data. Apart from supporting pre-allocated buffers it is
+	/// equivalent to to [`read()`](Self::read).
+	pub fn read_into(&self, buffer: &mut &mut [u8]) -> Result<()> {
+		let len = buffer.len();
+		let sliced = {
+			let buffer = core::mem::take(buffer);
```

Comment:
you take buffer so that an empty buffer is set in case of error, isn't it ? otherwise I think you can just do `sliced = &mut buffer[.....]`

Reply:
The main reason for the `take` is that is necessary in order to satisfy the borrow checker. I know of no other way to make this happen without resorting to unsafe code. Leaving behind an empty buffer on error is a nice extra.

Reply:
ok, but this compiles to me:
```diff
diff --git a/frame/contracts/src/chain_extension.rs b/frame/contracts/src/chain_extension.rs
index 662cfb2053..bd34bd4060 100644
--- a/frame/contracts/src/chain_extension.rs
+++ b/frame/contracts/src/chain_extension.rs
@@ -280,15 +280,11 @@ where
        /// equivalent to to [`read()`](Self::read).
        pub fn read_into(&self, buffer: &mut &mut [u8]) -> Result<()> {
                let len = buffer.len();
-               let sliced = {
-                       let buffer = core::mem::take(buffer);
-                       &mut buffer[..len.min(self.inner.input_len as usize)]
-               };
+               let sliced = &mut buffer[..len.min(self.inner.input_len as usize)];
                self.inner.runtime.read_sandbox_memory_into_buf(
                        self.inner.input_ptr,
                        sliced,
                )?;
-               *buffer = sliced;
                Ok(())
        }
 
```

Anyway its fine, I'm just surprised

Reply:
But this does not resize the slice that was passed in by the user, does it? This was the reason for the double reference.

Reply:
damn 🤦 yes